Libraries/DirectMLX.h

```diff
@@ -3503,7 +3503,17 @@ namespace dml
             totalElements *= inputTensorSizes[i];
         }
         TensorDesc outputCountTensor(DML_TENSOR_DATA_TYPE_UINT32, outputCountSizes, builder->GetTensorPolicy());
-        TensorDesc outputCoordinatesTensor(DML_TENSOR_DATA_TYPE_UINT32, {totalElements, dimensionCount}, builder->GetTensorPolicy());
+
+        if (!outputWidth)
+        {
+            outputWidth = dimensionCount;
```

Comment:
>outputWidth = dimensionCount; [](http://example.com/codeflow?start=12&length=29)

It looks functionally correct. Though, if the caller knows they added leading zeros, why not just have them not add them in the first place? DML hasn't had a requirement of 4D input for a while now. Alternately, if it is common that a particular caller has leading zeros in their tensors, is it limited to NonZeroCoordinates? Should we (as a corollary to Reinterpret) add a sort of `RemoveLeadingOnes` helper? I just don't want to end up seeing a number of other optional parameters added to other functions later (like OneHot or GatherNd...).

Btw, namingwise, it seems this isn't about any sort of spatial dimension "width" so much as tuple size (M x C), which originally had me quite confused 🙃.

Reply:
Only NonZeroCoordinates has this special behavior where having leading ones can change the size of the output. We could leave it that way and always force the last dimension of the output to be equal to the rank of the input, but ultimately dmlx is just supposed to be a wrapper around DirectML ops. You should be able to do the same thing with dmlx as you can do with the API directly.

Reply:
It could use some clarifying comments, because otherwise it's not so apparent the relationship between the input rank and this output coordinateIndicesPerTuple, and that this output size is really the `effective rank of the input`. Hmm, it might be clearer to just call it `inputEffectiveRank`, which is consistent in concept with DML's `InputDimensionCount` parameter to `GATHER_ND`.

```
struct DML_GATHER_ND_OPERATOR_DESC
{
    const DML_TENSOR_DESC* InputTensor;
    const DML_TENSOR_DESC* IndicesTensor;
    const DML_TENSOR_DESC* OutputTensor;
    UINT InputDimensionCount;
    UINT IndicesDimensionCount;
};
```

Though, I suppose you were also wanting to follow the precedent of passing an "output*" size parameter like other dmlx ops. 🤷‍♂️ Anyways, a comment would be nice.

Reply:
The more I think about it, the more I think it's better to leave it as-is :D. It might make it more confusing, and it's not like it's impossible to achieve what we want (it just makes it somewhat more convenient in some scenarios). I'll drop the PR for now as I don't want to start introducing patches like this unless it's absolutely necessary.